### ISSUE TITLE
Atlas size inreased on rebuild

### ISSUE THREAD

**Issue by mommy2914** [Elringus/SpriteDicing#9]
The sample asset (Animation.asset) is about 15MB.

However, an asset created with the same settings is about 8MB.
If I rebuild the asset, it will be about 15MB.

Is this a bug?

> SpriteDicing : v1.1
> Unity : 2019.4.5f1 / 2019.4.11f1

<img width="250" alt="Animation" src="https://user-images.githubusercontent.com/5304564/108693047-b762e300-7540-11eb-8fe3-71f3608c780c.png"><img width="250" alt="Animation_tmp" src="https://user-images.githubusercontent.com/5304564/108693088-c3e73b80-7540-11eb-9749-6c0391770cb4.png">

**Comment by Elringus** [comment]
It's just been generated with an old version of the package (size is cached).

**Comment by mommy2914** [comment]
Why is the size different the first time and the second time?
Animation_tmp.asset is 8MB the first time, but 15MB when rebuild from the inspector.

**Comment by mommy2914** [comment]
Left : Before build
Center : After build
Right : After rebuild

<img width="250" alt=" 2021-02-22 20 56 44" src="https://user-images.githubusercontent.com/5304564/108705675-f9942080-7550-11eb-8cbf-9af78cd088d9.png"><img width="250" alt=" 2021-02-22 20 57 56" src="https://user-images.githubusercontent.com/5304564/108705685-fc8f1100-7550-11eb-9d87-5306eef9bf09.png"><img width="250" alt=" 2021-02-22 20 59 01" src="https://user-images.githubusercontent.com/5304564/108705695-ff8a0180-7550-11eb-82cb-8803bfaf7ea2.png">

**Comment by Elringus** [comment]
Right, I can see the issue now. Thanks for sharing!

Here is the culprit: https://github.com/Elringus/SpriteDicing/blob/master/Assets/SpriteDicing/Editor/DicedSpriteAtlasEditor.cs#L553

[EditorUtility.CopySerialized](https://docs.unity3d.com/ScriptReference/EditorUtility.CopySerialized.html) is used to transfer sprite data from the newly generated sprites to the old ones contained in the atlas; that is required to preserve asset references. Seems like this method copies some excessive data.

If anyone have an idea how to preserve sprite references on rebuild without using CopySerialized, please let me know.

**Comment by mommy2914** [comment]
https://github.com/Elringus/SpriteDicing/blob/3d47cde94d6a6f60e92b097e9f707d59a91363c2/Assets/SpriteDicing/Editor/DicedSpriteAtlasEditor.cs#L202

I inserted the following code and found that m_RenderDataKey.second was 7950436238086893797. But when I looked at the asset file, m_RenderDataKey was 0.

The m_RenderDataKey in the asset file after rebuild is 7950436238086893797. The m_RenderDataKey of the file created by `AssetDatabase.CreateAsset(newSprite, tmpPath)` is 0.

Therefore, EditorUtility.CopySerialized doesn't copy some excessive data, and the loaded subassets are probably being transformed.


```
            var obj = new SerializedObject(spritesProperty.GetArrayElementAtIndex(0).objectReferenceValue);
            var sp = obj.GetIterator();
            var builder = new StringBuilder();
            while (sp.Next(true))
            {
                builder.AppendLine(sp.propertyPath);
                if (sp.propertyPath == "m_RenderDataKey.second")
                {
                    builder.AppendLine(sp.longValue.ToString());
                }
            }
            Debug.Log(builder.ToString());
```

**Comment by Elringus** [comment]
I'm pretty sure CopySerialized does add some excessive data; compare the initial and re-generated versions of the atlas asset in a VCS. m_RenderDataKey is not the only property being bloated after the copy, eg:

![](https://i.gyazo.com/10c583085a46339e246abf50a20d4e7a.png)

I've also found similar complaints about the CopySerialized on Unity forums, eg: https://forum.unity.com/threads/copyserialized-being-as-dumb-as-a-rock.276731

It's probably possible to manually iterate through the serialized properties of the sprites and copy only the required data, but I'd like to look for a simpler solution.

**Comment by mommy2914** [comment]
These are the properties I found to be bloated.
- m_RenderDataKey
- m_atlasRD
  - m_subMesh
    - indexCount
    - vertexCount
  - m_indexBuffer
  - m_vertexData
    - m_vertexCount
    - m_DataSize
      - m_typelessData
  - settingRaw

 Certainly, CopySerialized does add some excessive data.

I encountered a mysterious problem. I'm trying to implement create and build a DicedSpriteAtlas from a MenuItem.
Assets in the first build could become bloated. However, it does not always reproduce, and the cause has not been found.

```
        var target = AssetDatabase.LoadAssetAtPath<DicedSpriteAtlas>(atlasAssetPath);
        if (target == null)
        {
            target = ScriptableObject.CreateInstance<DicedSpriteAtlas>();
            AssetDatabase.CreateAsset(target, atlasAssetPath);
            AssetDatabase.SaveAssets();
        }

        // ↓ Same process as BuildAtlas
```

**Comment by Elringus** [comment]
We could as well try reporting bug to Unity. Will need a proper repro with just the sprite creation and copying, though.

**Comment by Elringus** [comment]
So I've attempted to make a repro for the bug report, but stumbled upon a regression: #10 Guess will have to wait until they fix it.

**Comment by Elringus** [comment]
I've submitted a bug report to Unity and they've created an issue on the tracker: [Diced sprite atlas is polluted with excessive data on rebuild](https://issuetracker.unity3d.com/product/unity/issues/guid/1327771) 

Please vote to make it processed as soon as possible.

**Comment by Elringus** [comment]
Fixed in #12

**Comment by Elringus** [comment]
It turned out (https://forum.unity.com/threads/1113004) that while there's no way to prevent the excessive data generation when modifying (loading) sprite assets, that data is actually editor-only and won't affect build size. To prevent inconsistent compression ratio reported on atlas rebuild, we'll now force editor-only data generation on initial build: https://github.com/Elringus/SpriteDicing/commit/5114dfcd6b1375513499798998505c31ae0fea7e
